### PR TITLE
[WIP] [SYSTEMML-540] Low-rank fully connected layer 

### PR DESCRIPTION
**Pull request by niketanpansare**
Please donot merge. This is work in-progress as this PR does not contain [expression for dU](https://github.com/niketanpansare/systemml/blob/37afb2df37e95a96488064f63fecc53311b84167/scripts/nn/layers/low_rank_affine.dml#L77).

This layer has three advantages over the affine layer:
1. It has significantly lower memory requirement than affine layer making it ideal for devices such as GPUs.
2. It implicitly avoids overfitting by minimizing the number of parameters in the neural network.
3. It can exploit sparsity-aware fused operators.

@prithvirajsen @bertholdreinwald @mboehm7 @dusenberrymw

### CODE REVIEW

File: scripts/nn/layers/low_rank_affine.dml [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+   *  - db: Gradient wrt `b`, of shape (1, M).
+   */
+  dX = dout %*% t(V) %*% t(U)
+  
+  # If out = Z %*% L, then dL = t(Z) %*% dout
+  # Substituting Z = X %*% U and L = V, we get
+  dV = t(U) %*% t(X) %*% dout
+    
+  # TODO:
+  dU = ??
```
Comment (dusenberrymw) [inline scripts/nn/layers/low_rank_affine.dml]: `dU = t(X) %*% dout %*% t(V)` should be correct

Comment (niketanpansare) [reply scripts/nn/layers/low_rank_affine.dml]: Yes, Sasha derived it independently and came to the same expression. I will add the gradient test in subsequent commit :+1:

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2245/<h2>Failed Tests: <span class='status-failure'>2</span></h2><h3><a name='SystemML-PullRequestBuilder/org.apache.systemml:systemml' /><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2245/org.apache.systemml$systemml/testReport'>SystemML-PullRequestBuilder/org.apache.systemml:systemml</a>: <span class='status-failure'>2</span></h3><ul><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2245/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.functions.jmlc/MulticlassSVMScoreTest/testJMLCMulticlassScoreSparseFlags/'><strong>org.apache.sysml.test.integration.functions.jmlc.MulticlassSVMScoreTest.testJMLCMulticlassScoreSparseFlags</strong></a></li><li><a href='https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2245/org.apache.systemml$systemml/testReport/org.apache.sysml.test.integration.scripts.nn/NNTest/testNNLibrary/'><strong>org.apache.sysml.test.integration.scripts.nn.NNTest.testNNLibrary</strong></a></li></ul>

**Comment by niketanpansare** [comment]
For comparisons, if I set `rank=10` for first fully connected layer, the memory budgets for weights are as follows:
```
+-----+---------------+--------------+--------------------+---------+-----------+---------+--------------------+
| Name|           Type|        Output|              Weight|     Bias|        Top|   Bottom|Memory* (train/test)|
+-----+---------------+--------------+--------------------+---------+-----------+---------+--------------------+
....
|  ip1|   InnerProduct| (, 512, 1, 1)|[3136 X 10], [10 ...|[1 X 512]|        ip1|    pool2|                20/2|
....
+-----+---------------+--------------+--------------------+---------+-----------+---------+--------------------+
```

instead of 
```
+-----+---------------+--------------+------------+---------+-----------+---------+--------------------+
| Name|           Type|        Output|      Weight|     Bias|        Top|   Bottom|Memory* (train/test)|
+-----+---------------+--------------+------------+---------+-----------+---------+--------------------+
....
|  ip1|   InnerProduct| (, 512, 1, 1)|[3136 X 512]|[1 X 512]|        ip1|    pool2|              797/13|
....
+-----+---------------+--------------+------------+---------+-----------+---------+--------------------+
```

The assumptions in the above numbers is that we avoid intermediate generation of `W` matrix in low-rank dml. The performance on the validation set is comparable to the numbers in [Caffe2DML's beginner's guide](http://apache.github.io/systemml/beginners-guide-caffe2dml) (but not the test set):
```
>>> lenet.fit(X_train, y_train)
Iter:100, training loss:0.1500353896492922, training accuracy:95.3125
Iter:200, training loss:0.11908411606279952, training accuracy:96.875
Iter:300, training loss:0.0360885446578709, training accuracy:98.4375
Iter:400, training loss:0.005613457983997603, training accuracy:100.0
Iter:500, training loss:0.007374364374341895, training accuracy:100.0
Iter:500, validation loss:252.23191312168126, validation accuracy:96.71875
Iter:600, training loss:0.0052617494034620455, training accuracy:100.0
Iter:700, training loss:7.714949309465524E-4, training accuracy:100.0
Iter:800, training loss:0.05139908897093824, training accuracy:98.4375
Iter:900, training loss:0.002646871516577846, training accuracy:100.0
Iter:1000, training loss:0.004223201373650876, training accuracy:100.0
Iter:1000, validation loss:240.54929655835912, validation accuracy:97.21354166666667
Iter:1100, training loss:0.0010717794786578477, training accuracy:100.0
Iter:1200, training loss:0.0035220468007343417, training accuracy:100.0
Iter:1300, training loss:5.602067121468543E-4, training accuracy:100.0
Iter:1400, training loss:0.0052189257916691704, training accuracy:100.0
Iter:1500, training loss:3.936150686732116E-4, training accuracy:100.0
Iter:1500, validation loss:237.41650599777304, validation accuracy:97.29166666666667
Iter:1600, training loss:0.0013395364335943797, training accuracy:100.0
Iter:1700, training loss:0.001646895497194339, training accuracy:100.0
Iter:1800, training loss:0.0017115306782601477, training accuracy:100.0
Iter:1900, training loss:2.1271571013680728E-4, training accuracy:100.0
Iter:2000, training loss:7.537731077508708E-4, training accuracy:100.0
Iter:2000, validation loss:241.23313586977062, validation accuracy:97.44791666666667
>>> lenet.score(X_test, y_test)
0.95599999999999996
```

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2249/

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/2251/
